File: frontend/src/types/book.ts
/**
 * TypeScript типы для книг.
 */

/** Формат файла книги */
export type FileFormat = 'txt' | 'epub' | 'fb2';

/** DTO для отображения книги */
export interface BookDTO {
  id: string; // UUID
  title: string;
  author: string | null;
  file_format: FileFormat;
  cover_thumbnail_path: string | null;
  date_added: string; // ISO datetime
}

/** DTO для списка книг с пагинацией */
export interface BookListDTO {
  books: BookDTO[];
  total: number;
  page?: number;
  page_size?: number;
}

/** Параметры для чанков книги */
export interface BookChunksResponse {
  book_id: string;
  chunks: BookChunkDTO[];
  total_chunks: number;
}

/** DTO для чанка книги */
export interface BookChunkDTO {
  id: string; // UUID чанка
  chunk_index: number;
  content_html: string;
  word_count: number;
}

/** Ответ с ошибкой от API */
export interface ErrorResponse {
  detail: string;
}

/** Параметры для получения списка книг */
export interface GetBooksParams {
  page?: number;
  page_size?: number;
  search?: string;
}

/** Данные для обновления книги */
export interface BookUpdate {
  title?: string;
  author?: string | undefined;
}

File: backend/src/schemas/book.py
"""
Pydantic схемы для книг.

Схемы:
- BookCreate: для загрузки книги
- BookUpdate: для обновления метаданных
- BookDTO: для отображения одной книги
- BookListDTO: для отображения списка книг с пагинацией
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, ConfigDict


class BookCreate(BaseModel):
    """Схема для загрузки книги.

    File upload обрабатывается через FastAPI UploadFile,
    поэтому эта схема пуста — валидация происходит в сервисе.
    """

    pass


class BookUpdate(BaseModel):
    """Схема для обновления метаданных книги."""

    title: Optional[str] = Field(None, max_length=500)
    author: Optional[str] = Field(None, max_length=300)


class BookDTO(BaseModel):
    """DTO для отображения книги."""

    id: UUID
    title: str
    author: Optional[str]
    file_format: str
    cover_thumbnail_path: Optional[str]
    date_added: datetime
    has_reading_position: bool

    model_config = ConfigDict(from_attributes=True)


class BookListDTO(BaseModel):
    """Схема для списка книг с пагинацией."""

    books: list[BookDTO]
    total: int
    page: int = 1
    page_size: int = 20


class BookChunkDTO(BaseModel):
    """DTO для отображения чанка книги."""

    chunk_index: int
    content_html: str
    word_count: int

    model_config = ConfigDict(from_attributes=True)


class BookChunksResponse(BaseModel):
    """Схема для ответа с чанками книги."""

    book_id: UUID
    chunks: list[BookChunkDTO]
    total_chunks: int


class ReadingPositionSave(BaseModel):
    """Схема для сохранения позиции чтения."""

    chunk_id: UUID
    offset: int = Field(ge=0)
    timestamp: datetime


class ReadingPositionResponse(BaseModel):
    """Схема для ответа с позицией чтения."""

    chunk_id: Optional[UUID] = None
    offset: int = 0
    timestamp: Optional[str] = None


class ReadingPositionSuccessResponse(BaseModel):
    """Схема для успешного сохранения позиции."""

    success: bool = True
